codec-dns/src/main/java/io/netty/handler/codec/dns/DnsMessage.java

```diff
@@ -15,225 +15,138 @@
+     * Sets the field Z. This field is reserved and should remain as 0 if the
+     * DNS server does not make usage of this field.
      *
-     * @param resource
-     *            the authority resource record to be added
-     * @return the message to allow method chaining
+     * @param z
+     *            the value for the reserved field Z
+     */
+    DnsMessage setZ(int z);
+
+    /**
+     * Returns the number of questions this message has.
      */
-    public DnsMessage addAuthorityResource(DnsResource resource) {
-        if (authority == null) {
-            authority = new LinkedList<DnsResource>();
-        }
-        authority.add(resource);
-        return this;
-    }
+    int questionCount();
 
     /**
-     * Adds an additional resource record to this message.
+     * Returns the number of answers this message has.
+     */
+    int answerCount();
+
+    /**
+     * Returns the number of authority records this message has.
+     */
+    int authorityRecordCount();
+
+    /**
+     * Returns the number of additional records this message has.
+     */
+    int additionalRecordCount();
+
+    /**
+     * Returns the first question of this message.
      *
-     * @param resource
-     *            the additional resource record to be added
-     * @return the message to allow method chaining
+     * @return {@code null} if this message doesn't have a question.
      */
-    public DnsMessage addAdditionalResource(DnsResource resource) {
-        if (additional == null) {
-            additional = new LinkedList<DnsResource>();
-        }
-        additional.add(resource);
-        return this;
-    }
+    DnsQuestion question();
 
-    @Override
-    protected void deallocate() {
-        // NOOP
-    }
+    /**
+     * Sets the question of this message.
+     */
+    DnsMessage setQuestion(DnsQuestion question);
 
-    @Override
-    public boolean release() {
-        release(questions());
-        release(answers());
-        release(additionalResources());
-        release(authorityResources());
-        return super.release();
-    }
-
-    private static void release(List<?> resources) {
-        for (Object resource: resources) {
-            ReferenceCountUtil.release(resource);
-        }
-    }
+    /**
+     * Returns the list of all the questions in this message.
+     */
+    List<DnsQuestion> questions();
 
-    @Override
-    public boolean release(int decrement) {
-        release(questions(), decrement);
-        release(answers(), decrement);
-        release(additionalResources(), decrement);
-        release(authorityResources(), decrement);
-        return super.release(decrement);
-    }
-
-    private static void release(List<?> resources, int decrement) {
-        for (Object resource: resources) {
-            ReferenceCountUtil.release(resource, decrement);
-        }
-    }
+    /**
+     * Returns the list of all the answer resource records in this message.
+     */
+    List<DnsRecord> answers();
```

Comment:
Consider returning `Interable<DnsRecord>` or having class implement that, unless `List` really does something useful here?